From 774d02a4370609fe9724325163de616ff2f7b1fa Mon Sep 17 00:00:00 2001
From: Erik Devriendt <erik.devriendt@siemens.com>
Date: Wed, 26 May 2010 13:40:57 +0200
Subject: [PATCH] Fix a crash when listing fonts on fortified builds
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

The code is writting the terminating NUL character past the panose
variable inside the TPFontName structure, which the fortify bound
checks catch and abort on. Even if the overwritten next variable
‘modifier’ gets initialized later on and thus there's no actual problem
with the code, it's obviously more correct and future-proof to only
write what the variable can hold.

Signed-off-by: Guillem Jover <guillem@hadrons.org>
---
 src/xfstt.cc |    7 ++++++-
 1 files changed, 6 insertions(+), 1 deletions(-)

diff --git a/src/xfstt.cc b/src/xfstt.cc
index 4eddd02..ead8b46 100644
--- a/src/xfstt.cc
+++ b/src/xfstt.cc
@@ -462,11 +462,16 @@ listTTFNFonts(char *pattern, int index, char *buf)
 	char *fontName = nameBase + ttfn->nameOfs;
 
 	TPFontName fn;
-	sprintf(&fn.panose[0][0], "%02X%02X%02X%02X%02X%02X%02X%02X%02X%02X",
+
+	/* Pre-allocate the buffer so that the terminating '\0' fits. */
+	char panose[sizeof(fn.panose) + 1];
+	sprintf(panose, "%02X%02X%02X%02X%02X%02X%02X%02X%02X%02X",
 		ttfn->bFamilyType, ttfn->bSerifStyle, ttfn->bWeight,
 		ttfn->bProportion, ttfn->bContrast, ttfn->bStrokeVariation,
 		ttfn->bArmStyle, ttfn->bLetterForm, ttfn->bMidLine,
 		ttfn->bXHeight);
+	/* Copy only the text, not the terminating '\0'. */
+	memcpy(&fn.panose[0][0], panose, sizeof(fn.panose));
 
 	fn.nameLen = sizeof(TPFontName) + ttfn->nameLen;
 	fn.magic[0] = 'T';
-- 
1.7.1

